timeboost-crypto/src/mre.rs

```diff
@@ -0,0 +1,220 @@
+//! Multi-recipient Encryption (MRE) allows encryption of a vector of message to all n parties.
+//! Proposed as MEGa in <https://eprint.iacr.org/2022/506>, this code implements the simplified
+//! variant in <https://eprint.iacr.org/2025/1175>.
+
+use ark_ec::{AffineRepr, CurveConfig, CurveGroup};
+use ark_serialize::serialize_to_vec;
+use ark_std::{
+    UniformRand,
+    rand::{CryptoRng, Rng},
+};
+use digest::Output;
+use serde::{Deserialize, Serialize};
+use sha2::Digest;
+use thiserror::Error;
+
+/// Ciphertext for multiple recipients in MRE scheme
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct MultiRecvCiphertext<C: CurveGroup, H: Digest> {
+    // the shared ephemeral public key (v:=g^beta in the paper)
+    epk: C::Affine,
+    // individual ciphertexts (e_i in the paper)
+    cts: Vec<Output<H>>,
+}
+
+impl<C: CurveGroup, H: Digest> MultiRecvCiphertext<C, H> {
+    /// Extract the recipient-specific ciphertext
+    pub fn get_recipient_ct(&self, index: usize) -> Option<SingleRecvCiphertext<C, H>> {
+        self.cts.get(index).map(|ct| SingleRecvCiphertext {
+            epk: self.epk,
+            ct: ct.clone(),
+        })
+    }
+}
+
+/// (Part of) [`MultiRecvCiphertext`] for a specific recipient.
+/// Only appropriate construction is [`MultiRecvCiphertext::get_recipient_ct()`]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct SingleRecvCiphertext<C: CurveGroup, H: Digest> {
+    pub(crate) epk: C::Affine,
+    pub(crate) ct: Output<H>,
+}
+
+/// Multi-Recipient Encryption (MRE) in <https://eprint.iacr.org/2025/1175>
+///
+/// # Parameters
+/// - `recipients` and `messages` must be of the same length and implicitly indexed
+/// - `aad` is the associated data
+/// - `C` is the DL group, `H` is the choice of H_enc whose output space = message space
+///   - preprocess messages to pad them to proper length before passing in
+pub fn encrypt<'pk, 'msg, C, H, R>(
+    recipients: impl IntoIterator<Item = &'pk C::Affine>,
+    messages: impl IntoIterator<Item = &'msg Vec<u8>>,
```

Comment:
curious about the advantages of explicit lifetimes compared to anonymous lifetimes here?

Reply:
```
 mre.rs    51  43 error           rustc [E0658]: anonymous lifetimes in `impl Trait` are unstable
 expected named lifetime parameter (eglot-check)
```

ah, don't think it will work

Reply:
My mistake. Was merely trying to point out that we can do lifetime elision here (but maybe there is a reason you want to keep them explicit).
```suggestion
    recipients: impl IntoIterator<Item = &C::Affine>,
    messages: impl IntoIterator<Item = &Vec<u8>>,
```

Reply:
your suggestion also won't work, you need to specify lifetime to pacify the compiler.